src/daft-io/src/google_cloud.rs

```diff
@@ -105,11 +134,22 @@ impl From<Error> for super::Error {
                 source: source.into(),
             },
             NotAFile { path } => Self::NotAFile { path },
+            UnableToGrabSemaphore { .. } => Self::Generic {
+                store: crate::SourceType::GCS,
+                source: error.into(),
+            },
+            UnableToCreateClient { .. } => Self::UnableToCreateClient {
+                store: crate::SourceType::GCS,
+                source: error.into(),
+            },
         }
     }
 }
 
-struct GCSClientWrapper(Client);
+struct GCSClientWrapper {
+    client: Client,
+    connection_pool_sema: Arc<Semaphore>,
```

Comment:
So IIUC this semaphore is:

1. Acquired when we initiate a connection to GCS
2. Released when the stream for that connection is exhausted?

For quick operations such as heads and stuff I guess we release it right after the result is obtained, hence the `_permit` pattern.

Could we add some short docstring here too describing that?